BUG: f2py appears to ignore Fortran compiler flags

### Describe the issue:

I'm collaborating on a code that compiles some fixed-form Fortran subroutines using `f2py`. My collaborators are using Python 3.11 and earlier, whereas I'm failing to build the subroutines on Python 3.12, presumably because I have to use the Meson backend (which I installed through my package manager).

Specifically, an example subroutine fails to build with `gfortran` without the `-ffixed-form` flag. But even if I pass this to `f2py`, I get the same error that `gfortran` gives me *without* the flag, as if it's not being passed.

I'm not 100% sure how all these form fields are going to render in the issue, but taking the Fortran source below as `mwe.f95`, I get (`$` represents the terminal prompt):
```
$ gfortran -c mwe.f95 
mwe.f95:1:22:

    1 |       subroutine test(
      |                      1
Error: Invalid character in name at (1)
mwe.f95:2:7:

    2 |      & x)
      |       1
Error: Invalid character in name at (1)
mwe.f95:4:9:

    4 |       end subroutine test
      |         1
Error: Expecting END PROGRAM statement at (1)
f951: Error: Unexpected end of file in ‘mwe.f95’
$ gfortran -ffixed-form -c mwe.f95  # no output
$ f2py -c --f77flags=-ffixed-form --f90flags=-ffixed-form -m mwe mwe.f95
< full error in the appropriate field but relevant detail is... >
...

[3/6] Compiling Fortran object mwe.cpython-312-x86_64-linux-gnu.so.p/mwe.f95.o
FAILED: mwe.cpython-312-x86_64-linux-gnu.so.p/mwe.f95.o 
gfortran -Imwe.cpython-312-x86_64-linux-gnu.so.p -I. -I.. -I/home/wball/.local/lib/python3.12/site-packages/numpy/core/include -I/home/wball/.local/lib/python3.12/site-packages/numpy/f2py/src -I/usr/include/python3.12 -fvisibility=hidden -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -O3 -fPIC -Jmwe.cpython-312-x86_64-linux-gnu.so.p -o mwe.cpython-312-x86_64-linux-gnu.so.p/mwe.f95.o -c ../mwe.f95
../mwe.f95:1:22:

    1 |       subroutine test(
      |                      1
Error: Invalid character in name at (1)
../mwe.f95:2:7:

    2 |      & x)
      |       1
Error: Invalid character in name at (1)
../mwe.f95:4:9:

    4 |       end subroutine test
      |         1
Error: Expecting END PROGRAM statement at (1)
f951: Error: Unexpected end of file in ‘../mwe.f95’
...
subprocess.CalledProcessError: Command '['meson', 'compile', '-C', 'bbdir']' returned non-zero exit status 1.
```
Note that the build succeeds if I change the source file extension to `.f`, perhaps because the fixed-form is inferred.

### Reproduce the code example:

```python
! mwe.f95
      subroutine test(
     & x)
          real*8 x
      end subroutine test
```


### Error message:

```shell
Cannot use distutils backend with Python>=3.12, using meson backend instead.
Using meson backend
Will pass --lower to f2py
See https://numpy.org/doc/stable/f2py/buildtools/meson.html
Reading fortran codes...
	Reading file 'mwe.f95' (format:free)
Line #1 in mwe.f95:"      subroutine test( "
	analyzeline: No name/args pattern found for line.
Post-processing...
	Block: mwe
			Block: unknown_subroutine
Applying post-processing hooks...
  character_backward_compatibility_hook
Post-processing (stage 2)...
Building modules...
    Building module "mwe"...
    Generating possibly empty wrappers"
    Maybe empty "mwe-f2pywrappers.f"
        Constructing wrapper function "unknown_subroutine"...
          unknown_subroutine()
    Wrote C/API module "mwe" to file "./mwemodule.c"
The Meson build system
Version: 1.3.1
Source dir: /tmp/tmpjodywqy7
Build dir: /tmp/tmpjodywqy7/bbdir
Build type: native build
Project name: mwe
Project version: 0.1
Fortran compiler for the host machine: gfortran (gcc 13.2.1 "GNU Fortran (GCC) 13.2.1 20231205 (Red Hat 13.2.1-6)")
Fortran linker for the host machine: gfortran ld.bfd 2.40-14
C compiler for the host machine: cc (gcc 13.2.1 "cc (GCC) 13.2.1 20231205 (Red Hat 13.2.1-6)")
C linker for the host machine: cc ld.bfd 2.40-14
Host machine cpu family: x86_64
Host machine cpu: x86_64
Program /usr/bin/python3 found: YES (/usr/bin/python3)
Found pkg-config: YES (/usr/bin/pkg-config) 1.9.5
Run-time dependency python found: YES 3.12
Library quadmath found: YES
Build targets in project: 1

Found ninja-1.11.1 at /usr/bin/ninja
INFO: autodetecting backend as ninja                                                                                                                                                          
INFO: calculating backend command to run: /usr/bin/ninja -C /tmp/tmpjodywqy7/bbdir
ninja: Entering directory `/tmp/tmpjodywqy7/bbdir'
[3/6] Compiling Fortran object mwe.cpython-312-x86_64-linux-gnu.so.p/mwe.f95.o
FAILED: mwe.cpython-312-x86_64-linux-gnu.so.p/mwe.f95.o 
gfortran -Imwe.cpython-312-x86_64-linux-gnu.so.p -I. -I.. -I/home/wball/.local/lib/python3.12/site-packages/numpy/core/include -I/home/wball/.local/lib/python3.12/site-packages/numpy/f2py/src -I/usr/include/python3.12 -fvisibility=hidden -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -O3 -fPIC -Jmwe.cpython-312-x86_64-linux-gnu.so.p -o mwe.cpython-312-x86_64-linux-gnu.so.p/mwe.f95.o -c ../mwe.f95
../mwe.f95:1:22:

    1 |       subroutine test(
      |                      1
Error: Invalid character in name at (1)
../mwe.f95:2:7:

    2 |      & x)
      |       1
Error: Invalid character in name at (1)
../mwe.f95:4:9:

    4 |       end subroutine test
      |         1
Error: Expecting END PROGRAM statement at (1)
f951: Error: Unexpected end of file in ‘../mwe.f95’
[5/6] Compiling C object mwe.cpython-312-x86_64-linux-gnu.so.p/15b943306aebd8c2a7f428d911374b526ccafd56_.._.._f2py_src_fortranobject.c.o
ninja: build stopped: subcommand failed.
Traceback (most recent call last):
  File "/usr/bin/f2py", line 33, in <module>
    sys.exit(load_entry_point('numpy==1.24.4', 'console_scripts', 'f2py')())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/wball/.local/lib/python3.12/site-packages/numpy/f2py/f2py2e.py", line 766, in main
    run_compile()
  File "/home/wball/.local/lib/python3.12/site-packages/numpy/f2py/f2py2e.py", line 738, in run_compile
    builder.compile()
  File "/home/wball/.local/lib/python3.12/site-packages/numpy/f2py/_backends/_meson.py", line 178, in compile
    self.run_meson(self.build_dir)
  File "/home/wball/.local/lib/python3.12/site-packages/numpy/f2py/_backends/_meson.py", line 173, in run_meson
    self._run_subprocess_command(compile_command, build_dir)
  File "/home/wball/.local/lib/python3.12/site-packages/numpy/f2py/_backends/_meson.py", line 167, in _run_subprocess_command
    subprocess.run(command, cwd=cwd, check=True)
  File "/usr/lib64/python3.12/subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['meson', 'compile', '-C', 'bbdir']' returned non-zero exit status 1.
```


### Python and NumPy Versions:

1.26.3
3.12.1 (main, Dec 18 2023, 00:00:00) [GCC 13.2.1 20231205 (Red Hat 13.2.1-6)]


### Runtime Environment:

[{'numpy_version': '1.26.3',
  'python': '3.12.1 (main, Dec 18 2023, 00:00:00) [GCC 13.2.1 20231205 (Red '
            'Hat 13.2.1-6)]',
  'uname': uname_result(system='Linux', node='localhost.localdomain', release='6.7.3-200.fc39.x86_64', version='#1 SMP PREEMPT_DYNAMIC Thu Feb  1 03:29:52 UTC 2024', machine='x86_64')},
 {'simd_extensions': {'baseline': ['SSE', 'SSE2', 'SSE3'],
                      'found': ['SSSE3',
                                'SSE41',
                                'POPCNT',
                                'SSE42',
                                'AVX',
                                'F16C',
                                'FMA3',
                                'AVX2'],
                      'not_found': ['AVX512F',
                                    'AVX512CD',
                                    'AVX512_KNL',
                                    'AVX512_KNM',
                                    'AVX512_SKX',
                                    'AVX512_CLX',
                                    'AVX512_CNL',
                                    'AVX512_ICL']}},
 {'architecture': 'Haswell',
  'filepath': '/home/wball/.local/lib/python3.12/site-packages/numpy.libs/libopenblas64_p-r0-0cf96a72.3.23.dev.so',
  'internal_api': 'openblas',
  'num_threads': 4,
  'prefix': 'libopenblas',
  'threading_layer': 'pthreads',
  'user_api': 'blas',
  'version': '0.3.23.dev'},
 {'filepath': '/usr/lib64/libgomp.so.1.0.0',
  'internal_api': 'openmp',
  'num_threads': 4,
  'prefix': 'libgomp',
  'user_api': 'openmp',
  'version': None}]
None


### Context for the issue:

This appears to be a regression in `f2py`, in that something that was working under defaults in Python 3.11 (with distutils) appears to broken in Python 3.12 (with Meson). I suspect this is actually a Meson issue but I have limited experience with it.

I have access to a few different systems to try a few tests. If the intention isn't to support what I'm trying, it should be documented. I couldn't (at least in the ~20 minutes I spent paging through) find anything relevant in the [f2py docs](https://numpy.org/doc/stable/f2py/index.html).

--- a/numpy/f2py/_backends/_meson.py
+++ b/numpy/f2py/_backends/_meson.py
@@ -223,11 +223,11 @@ def _prepare_sources(mname, sources, bdir):
 def _get_flags(fc_flags):
     flag_values = []
     flag_pattern = re.compile(r"--f(77|90)flags=(.*)")
-
     for flag in fc_flags:
         match_result = flag_pattern.match(flag)
         if match_result:
             values = match_result.group(2).strip().split()
+            values = [val.strip("'\"") for val in values]
             flag_values.extend(values)
     # Hacky way to preserve order of flags
     unique_flags = list(dict.fromkeys(flag_values))

--- a/numpy/f2py/_backends/_meson.py
+++ b/numpy/f2py/_backends/_meson.py
@@ -28,7 +28,7 @@ class MesonTemplate:
         include_dirs: list[Path],
         object_files: list[Path],
         linker_args: list[str],
-        c_args: list[str],
+        fortran_args: list[str],
         build_type: str,
         python_exe: str,
     ):
@@ -46,12 +46,18 @@ class MesonTemplate:
             self.include_dirs = []
         self.substitutions = {}
         self.objects = object_files
+        # Convert args to '' wrapped variant for meson
+        self.fortran_args = [
+            f"'{x}'" if not (x.startswith("'") and x.endswith("'")) else x
+            for x in fortran_args
+        ]
         self.pipeline = [
             self.initialize_template,
             self.sources_substitution,
             self.deps_substitution,
             self.include_substitution,
             self.libraries_substitution,
+            self.fortran_args_substitution,
         ]
         self.build_type = build_type
         self.python_exe = python_exe
@@ -109,6 +115,14 @@ class MesonTemplate:
             [f"{self.indent}'''{inc}'''," for inc in self.include_dirs]
         )
 
+    def fortran_args_substitution(self) -> None:
+        if self.fortran_args:
+            self.substitutions["fortran_args"] = (
+                f"{self.indent}fortran_args: [{', '.join([arg for arg in self.fortran_args])}],"
+            )
+        else:
+            self.substitutions["fortran_args"] = ""
+
     def generate_meson_build(self):
         for node in self.pipeline:
             node()
@@ -126,6 +140,7 @@ class MesonBackend(Backend):
         self.build_type = (
             "debug" if any("debug" in flag for flag in self.fc_flags) else "release"
         )
+        self.fc_flags = _get_flags(self.fc_flags)
 
     def _move_exec_to_root(self, build_dir: Path):
         walk_dir = Path(build_dir) / self.meson_build_dir
@@ -203,3 +218,17 @@ def _prepare_sources(mname, sources, bdir):
         if not Path(source).suffix == ".pyf"
     ]
     return extended_sources
+
+
+def _get_flags(fc_flags):
+    flag_values = []
+    flag_pattern = re.compile(r"--f(77|90)flags=(.*)")
+
+    for flag in fc_flags:
+        match_result = flag_pattern.match(flag)
+        if match_result:
+            values = match_result.group(2).strip().split()
+            flag_values.extend(values)
+    # Hacky way to preserve order of flags
+    unique_flags = list(dict.fromkeys(flag_values))
+    return unique_flags

--- a/numpy/f2py/_backends/meson.build.template
+++ b/numpy/f2py/_backends/meson.build.template
@@ -51,4 +51,5 @@ ${dep_list}
 ${lib_list}
 ${lib_dir_list}
                      ],
+${fortran_args}
                      install : true)

--- a/numpy/f2py/f2py2e.py
+++ b/numpy/f2py/f2py2e.py
@@ -635,6 +635,7 @@ def run_compile():
         r'--((f(90)?compiler(-exec|)|compiler)=|help-compiler)')
     flib_flags = [_m for _m in sys.argv[1:] if _reg3.match(_m)]
     sys.argv = [_m for _m in sys.argv if _m not in flib_flags]
+    # TODO: Once distutils is dropped completely, i.e. min_ver >= 3.12, unify into --fflags
     reg_f77_f90_flags = re.compile(r'--f(77|90)flags=')
     reg_distutils_flags = re.compile(r'--((f(77|90)exec|opt|arch)=|(debug|noopt|noarch|help-fcompiler))')
     fc_flags = [_m for _m in sys.argv[1:] if reg_f77_f90_flags.match(_m)]

--- a/numpy/f2py/f2py2e.py
+++ b/numpy/f2py/f2py2e.py
@@ -635,10 +635,13 @@ def run_compile():
         r'--((f(90)?compiler(-exec|)|compiler)=|help-compiler)')
     flib_flags = [_m for _m in sys.argv[1:] if _reg3.match(_m)]
     sys.argv = [_m for _m in sys.argv if _m not in flib_flags]
-    _reg4 = re.compile(
-        r'--((f(77|90)(flags|exec)|opt|arch)=|(debug|noopt|noarch|help-fcompiler))')
-    fc_flags = [_m for _m in sys.argv[1:] if _reg4.match(_m)]
-    sys.argv = [_m for _m in sys.argv if _m not in fc_flags]
+    reg_f77_f90_flags = re.compile(r'--f(77|90)flags=')
+    reg_distutils_flags = re.compile(r'--((f(77|90)exec|opt|arch)=|(debug|noopt|noarch|help-fcompiler))')
+    fc_flags = [_m for _m in sys.argv[1:] if reg_f77_f90_flags.match(_m)]
+    distutils_flags = [_m for _m in sys.argv[1:] if reg_distutils_flags.match(_m)]
+    if not (MESON_ONLY_VER or backend_key == 'meson'):
+        fc_flags.extend(distutils_flags)
+    sys.argv = [_m for _m in sys.argv if _m not in (fc_flags + distutils_flags)]
 
     del_list = []
     for s in flib_flags:

--- a/numpy/f2py/tests/src/regression/f77fixedform.f95
+++ b/numpy/f2py/tests/src/regression/f77fixedform.f95
@@ -1,0 +1,5 @@
+C This is an invalid file, but it does compile with -ffixed-form
+      subroutine mwe(
+     & x)
+          real x
+      end subroutine mwe

--- a/numpy/f2py/tests/test_regression.py
+++ b/numpy/f2py/tests/test_regression.py
@@ -130,8 +130,8 @@ def test_gh25784():
             [util.getpath("tests", "src", "regression", "f77fixedform.f95")],
             options=[
                 # Meson will collect and dedup these to pass to fortran_args:
-                "--f77flags='-ffixed-form'",
-                "--f90flags='-ffixed-form'",
+                "--f77flags='-ffixed-form -O2'",
+                "--f90flags=\"-ffixed-form -Og\"",
             ],
             module_name="Blah",
         )

--- a/numpy/f2py/tests/test_regression.py
+++ b/numpy/f2py/tests/test_regression.py
@@ -109,6 +109,7 @@ class TestF90Contiuation(util.F2PyTest):
         assert(res[0] == 8)
         assert(res[1] == 15)
 
+@pytest.mark.slow
 def test_gh26623():
     # Including libraries with . should not generate an incorrect meson.build
     try:
@@ -119,3 +120,20 @@ def test_gh26623():
         )
     except RuntimeError as rerr:
         assert "lparen got assign" not in str(rerr)
+
+
+@pytest.mark.slow
+def test_gh25784():
+    # Compile dubious file using passed flags
+    try:
+        aa = util.build_module(
+            [util.getpath("tests", "src", "regression", "f77fixedform.f95")],
+            options=[
+                # Meson will collect and dedup these to pass to fortran_args:
+                "--f77flags='-ffixed-form'",
+                "--f90flags='-ffixed-form'",
+            ],
+            module_name="Blah",
+        )
+    except ImportError as rerr:
+        assert "unknown_subroutine_" in str(rerr)

--- a/numpy/f2py/tests/util.py
+++ b/numpy/f2py/tests/util.py
@@ -121,7 +121,7 @@ def build_module(source_files, options=[], skip=[], only=[], module_name=None):
         dst_sources.append(dst)
 
         base, ext = os.path.splitext(dst)
-        if ext in (".f90", ".f", ".c", ".pyf"):
+        if ext in (".f90", ".f95", ".f", ".c", ".pyf"):
             f2py_sources.append(dst)
 
     assert f2py_sources

